File: C# (server)/ClientServerWithJava/Networking/ShowServicesRpcProxy.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading.Tasks;
using Service;
using Model;

namespace Networking
{
    public class ShowServicesRpcProxy : IService
    {
        private string host;
        private int port;

        private Observer client;
        public BinaryFormatter binaryFormatter { get; set; }
        private TcpClient connection;
        public NetworkStream networkStream { get; set; }
        private volatile bool finished;
        private BlockingCollection<Response> queueResponses;

        public ShowServicesRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            binaryFormatter = new BinaryFormatter();
            queueResponses = new BlockingCollection<Response>();
        }

        public User Login(string username, string password, Observer client)
        {
            InitializeConnection();
            UserDTO userDTO = new UserDTO(username, password);
            Request request = new Request.Builder().Type(RequestType.LOGIN).Data(userDTO).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.OK)
            {
                this.client = client;
                UserDTO receivedUserDTO = (UserDTO)response.Data;
                User user = new User(receivedUserDTO.Name, receivedUserDTO.Password);
                return user;
            }
            if (response.Type == ResponseType.ERROR)
            {
                string error = response.Data.ToString();
                CloseConnection();
                throw new ArgumentException(error);
            }
            return null;
        }

        public void Logout(string username)
        {
            Request request = new Request.Builder().Type(RequestType.LOGOUT).Data(username).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string error = response.Data.ToString();
                throw new ArgumentException(error);
            }
        }

        public MatchDTO GetMatch(int i)
        {
            Request request = new Request.Builder().Type(RequestType.GET_MATCH).Data(i).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.GET_MATCH)
            {
                return (MatchDTO)response.Data;
            }
            if (response.Type == ResponseType.ERROR)
            {
                string error = response.Data.ToString();
                throw new ArgumentException(error);
            }
            return null;
        }

        public bool AddTicket(BuyTicketsDTO ticket)
        {
            Request request = new Request.Builder().Type(RequestType.BUY_TICKETS).Data(ticket).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string error = response.Data.ToString();
                throw new ArgumentException(error);
            }
            return true;
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                connection.Close();
                client = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
        }

        private void SendRequest(Request request)
        {
            try
            {
                MemoryStream memoryStream = new MemoryStream();
                binaryFormatter.Serialize(memoryStream, request);
                networkStream.Write(memoryStream.GetBuffer(), 0, (int)memoryStream.Length);
                networkStream.Flush();
            }
            catch (Exception e)
            {
                throw new ArgumentException("Error sending object " + e);
            }
        }

        private Response ReadResponse()
        {
            Response response = null;
            try
            {
                networkStream = connection.GetStream();
                byte[] buffer = new byte[1024]; // Adjust buffer size as needed
                int bytesRead = networkStream.Read(buffer, 0, buffer.Length);
                MemoryStream receivedStream = new MemoryStream(buffer);
                response = (Response)binaryFormatter.Deserialize(receivedStream);
                networkStream.Flush();
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception);
            }
            return response;
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient();
                connection.Connect(host, port);

                networkStream = connection.GetStream();
        
                finished = false;
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
        }

        private void StartReader()
        {
            Task tw = Task.Run(() => ReaderThread.Run(this));
        }

        public ICollection<string> GetListMatches(string search)
        {
            Request request = new Request.Builder().Type(RequestType.GET_MATCHES).Data(search).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string error = response.Data.ToString();
                throw new ArgumentException(error);
            }
            return (ICollection<string>)response.Data;
        }

        private void HandleUpdate(Response response)
        {
            if (response.Type == ResponseType.UPDATE)
            {
                client.Update();
            }
        }

        private bool IsUpdateResponse(Response response)
        {
            return response.Type == ResponseType.UPDATE;
        }

        private class ReaderThread
        {
            public static void Run(ShowServicesRpcProxy proxy)
            {
                while (!proxy.finished)
                {
                    try
                    {
                            proxy.networkStream = proxy.connection.GetStream();
                            byte[] buffer = new byte[1024]; // Adjust buffer size as needed
                            int bytesRead = proxy.networkStream.Read(buffer, 0, buffer.Length);
                            MemoryStream receivedStream = new MemoryStream(buffer);
                            Response receivedResponse = (Response)proxy.binaryFormatter.Deserialize(receivedStream);
                            //object obj = proxy.binaryFormatter.Deserialize(proxy.networkStream);
                            //Response receivedResponse = (Response)obj;
                            Console.WriteLine("Response received " + receivedResponse);
                            if (proxy.IsUpdateResponse(receivedResponse))
                            {
                                proxy.HandleUpdate(receivedResponse);
                            }
                            else
                            {
                                try
                                {
                                    proxy.queueResponses.Add(receivedResponse);
                                }
                                catch (Exception e)
                                {
                                    Console.WriteLine(e);
                                }
                            }
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine("Reading error " + e);
                    }
                }
            }
        }
    }
}

File: Java (client)/Networking/src/main/java/ro/mpp2024/ShowServicesRpcProxy.java
package ro.mpp2024;

import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.Collection;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ShowServicesRpcProxy implements IService{
    private String host;
    private int port;

    private Observer client;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private volatile boolean finished;
    private BlockingQueue<Response> queueResponses;

    public ShowServicesRpcProxy(String host, int port) {
        this.host = host;
        this.port = port;
        queueResponses = new LinkedBlockingQueue<>();
    }

    @Override
    public Optional<User> login(String username, String password, Observer client) {
        initializeConnection();
        UserDTO utilizator = new UserDTO(username, password);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(utilizator).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.OK) {
            this.client = client;
            UserDTO userDTO = (UserDTO) response.getData();
            User user = new User(userDTO.getName(), userDTO.getPassword());
            return Optional.of(user);
        }
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            closeConnection();
            throw new IllegalArgumentException(error);
        }
        return Optional.empty();
    }

    @Override
    public void logout(String username) {
        Request request = new Request.Builder().type(RequestType.LOGOUT).data(username).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new IllegalArgumentException(error);
        }
    }

    @Override
    public MatchDTO getMatch(int i) {
        Request request = new Request.Builder().type(RequestType.GET_MATCH).data(i).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.GET_MATCH) {
            return (MatchDTO) response.getData();
        }
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new IllegalArgumentException(error);
        }
        return null;
    }

    @Override
    public boolean addTicket(BuyTicketsDTO ticket) {
        Request request = new Request.Builder().type(RequestType.BUY_TICKETS).data(ticket).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new IllegalArgumentException(error);
        }
        return true;
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) {
        try {
            output.writeObject(request);
            output.flush();
        } catch (Exception e) {
            throw new IllegalArgumentException("Error sending object " + e);
        }
    }

    private Response readResponse() {
        Response response = null;
        try {
            response = queueResponses.take();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }

    private void initializeConnection() {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }
    @Override
    public synchronized Collection<String> getListMatches(String search) {
        Request request = new Request.Builder().type(RequestType.GET_MATCHES).data(search).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new IllegalArgumentException(error);
        }
        return (Collection<String>) response.getData();
    }

    private void handleUpdate(Response response) {
        if (response.getType() == ResponseType.UPDATE) {
            client.update();
        }
    }
    private boolean isUpdateResponse(Response response) {
        return response.getType() == ResponseType.UPDATE;
    }

    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    System.out.println("Response received " + response);
                    if (response instanceof Response) {
                        Response response1 = (Response) response;
                        if (isUpdateResponse(response1)) {
                            handleUpdate(response1);
                        } else {
                            try {
                                queueResponses.put((Response) response);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Reading error " + e);
                }
            }
        }
    }
}
